### PR TITLE
security: tikv encryption kms config

### PR DESCRIPTION
**Pull request by shuijing198799**
### What problem does this PR solve? 
Add static encrytion of tikv data config in tikv config 
### What is changed and how does it work?
Add static encrytion of tikv data config in tikv config , generate the openapi, crd and docs
### Check List 

Tests 

 - E2E test

Code changes

 - Has Go code change

```release-note
add filed `TiKVEncryptionConfig` struct in `TiKVConfig` , it defines how to encrytion data key and raw data in TiKV, and what backup store the master key, details see the descryption in tikv_config.go
```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/tikv_config.go [hunk 1]
```diff
@@ -742,3 +744,72 @@ type TiKVCoprocessorConfig struct {
 	// optional
 	RegionSplitKeys *int64 `json:"region-split-keys,omitempty" toml:"region-split-keys,omitempty"`
 }
+
+type TiKVEncryptionConfig struct {
```
Comment (aylei) [inline pkg/apis/pingcap/v1alpha1/tikv_config.go:748]: add `// +k8s:openapi-gen=true` to generate open api spec of these structs

File: pkg/apis/pingcap/v1alpha1/tikv_config.go [hunk 2]
```diff
@@ -742,3 +744,72 @@ type TiKVCoprocessorConfig struct {
...
+
+	// Master key config
+	MasterKey *TiKVMasterKeyConfig `json:"master-key,omitempty" toml:"master-key,omitempty"`
+
+	// Previous master key config
+	// It used in master key rotation, the data key should decryption by previous master key and  then encrypytion by new master key
+	PreviousMasterKey *TiKVMasterKeyConfig `json:"previous-master-key,omitempty" toml:"previoud-master-key,omitempty"`
+}
+
+type TiKVMasterKeyConfig struct {
```
Comment (aylei) [inline pkg/apis/pingcap/v1alpha1/tikv_config.go:768]: ditto

File: pkg/apis/pingcap/v1alpha1/tikv_config.go [hunk 3]
```diff
@@ -742,3 +744,72 @@ type TiKVCoprocessorConfig struct {
...
+	// Master key file config
+	// If the type set to file, this config should be filled
+	MasterKeyFileConfig `json:",inline"`
+
+	// Master key KMS config
+	// If the type set to kms, this config should be filled
+	MasterKeyKMSConfig `json:",inline"`
+}
+
+type MasterKeyFileConfig struct {
```
Comment (aylei) [inline pkg/apis/pingcap/v1alpha1/tikv_config.go:783]: ditto

File: pkg/apis/pingcap/v1alpha1/tikv_config.go [hunk 4]
```diff
@@ -742,3 +744,72 @@ type TiKVCoprocessorConfig struct {
...
+	// Text file containing the key in hex form, end with '\n'
+	Path string `json:"path" toml:"path"`
+}
+
+type MasterKeyKMSConfig struct {
+	// AWS CMK key-id it can be find in AWS Console or use aws cli
+	// This field is required
+	KeyID string `json:"key-id" toml:"key-id"`
+
+	// AccessKey of AWS user, leave empty if using other authrization method
```
Comment (aylei) [inline pkg/apis/pingcap/v1alpha1/tikv_config.go:799]: ditto

**Comment by weekface** [comment]
what is the [**document**](https://pingcap.com/docs/stable/) of this encryption feature? @shuijing198799

File: pkg/apis/pingcap/v1alpha1/tikv_config.go (hunk 1 above)
Comment (shuijing198799) [reply pkg/apis/pingcap/v1alpha1/tikv_config.go:749]: done

File: pkg/apis/pingcap/v1alpha1/tikv_config.go (hunk 2 above)
Comment (shuijing198799) [reply pkg/apis/pingcap/v1alpha1/tikv_config.go:770]: done

File: pkg/apis/pingcap/v1alpha1/tikv_config.go (hunk 3 above)
Comment (shuijing198799) [reply pkg/apis/pingcap/v1alpha1/tikv_config.go:786]: done

File: pkg/apis/pingcap/v1alpha1/tikv_config.go (hunk 4 above)
Comment (shuijing198799) [reply pkg/apis/pingcap/v1alpha1/tikv_config.go:803]: done

**Comment by shuijing198799** [comment]
> what is the [**document**](https://pingcap.com/docs/stable/) of this encryption feature? @shuijing198799

this feature is compeleted by tikv few days ago, we have no doc now.

File: pkg/apis/pingcap/v1alpha1/tikv_config.go [hunk 5]
```diff
@@ -742,3 +744,76 @@ type TiKVCoprocessorConfig struct {
...
+	// Encrypyion method, use data key encryption raw rocksdb data
+	// Possible values: plaintext, aes128-ctr, aes192-ctr, aes256-ctr
+	// Optional: Default to plaintext
+	// optional
+	Method string `json:"method,omitempty" toml:"method,omitempty"`
+
+	// The frequency of datakey rotation, It managered by tikv
+	// Optional: default to 7d
+	// optional
+	DataKeyRotationPeriod string `json:"data-key-rotation-period,omitempty" toml:"data-key-rotation-period,omitempty"`
```
Comment (Yisaer) [inline pkg/apis/pingcap/v1alpha1/tikv_config.go:759]: \*string

File: pkg/apis/pingcap/v1alpha1/tikv_config.go [hunk 6]
```diff
@@ -742,3 +744,76 @@ type TiKVCoprocessorConfig struct {
...
 	RegionSplitKeys *int64 `json:"region-split-keys,omitempty" toml:"region-split-keys,omitempty"`
 }
+
+// +k8s:openapi-gen=true
+type TiKVEncryptionConfig struct {
+	// Encrypyion method, use data key encryption raw rocksdb data
+	// Possible values: plaintext, aes128-ctr, aes192-ctr, aes256-ctr
+	// Optional: Default to plaintext
+	// optional
+	Method string `json:"method,omitempty" toml:"method,omitempty"`
```
Comment (Yisaer) [inline pkg/apis/pingcap/v1alpha1/tikv_config.go:754]: \*string

File: pkg/apis/pingcap/v1alpha1/tikv_config.go [hunk 7]
```diff
@@ -742,3 +744,76 @@ type TiKVCoprocessorConfig struct {
...
+	SecretKey string `json:"secret-access-key,omitempty" toml:"access-key,omitempty"`
+
+	// Region of this KMS key
+	// Optional: Default to us-east-1
+	// optional
+	Region string `json:"region,omitempty" toml:"region,omitempty"`
+
+	// Used for KMS compatible KMS, such as Ceph, minio, If use AWS, leave empty
+	// optional
+	Endpoint string `json:"endpoint,omitempty" toml:"endpoint,omitempty"`
```
Comment (Yisaer) [inline pkg/apis/pingcap/v1alpha1/tikv_config.go:818]: If the these properties could be omitted, maybe we should define them as \*string?

Comment (shuijing198799) [reply pkg/apis/pingcap/v1alpha1/tikv_config.go:818]: I don't think it is neccessary， maybe you think this is friendly for frontend, but  operator will not face to front end.

**Pull request merged**